File: Rust/interview_preparation_kit/problem_2_counting_valleys.rs
use std::time::Instant;

fn counting_valleys(steps: i32, path: &str) -> i32 {
    let mut level = 0;
    let mut valley_count = 0;
    for (i, step) in path.chars().enumerate() {
        if i as i32 >= steps {
            break;
        }

        if step == 'U' {
            level += 1;
            if level == 0 {
                valley_count += 1;
            }
        } else {
            level -= 1;
        }
    }

    valley_count
}

fn main() {
    let steps = 8;
    let path = "UDDDUDUU";
    let start = Instant::now();
    let result = counting_valleys(steps, path);
    println!("Result: {}", result);
    println!("Elapsed time: {:?}", start.elapsed());
}

File: C/interview_preparation_kit/general/problem_2_counting_valleys.c
#include <stdio.h>
#include <time.h>

int countingValleys(int steps, const char *path) {
  int valleys = 0;
  int level = 0;

  for (int i = 0; i < steps; i++)
  {
    if (path[i] == 'U')
    {
      level++;
      if (level == 0)
        valleys++;
    }
    else
    {
      level--;
    }
  }

  return valleys;
}

int main()
{
  int steps = 8;
  const char *path = "UDDDUDUU";

  clock_t start = clock();
  int result = countingValleys(steps, path);
  clock_t end = clock();
  double execution_time = (double)(end - start) / CLOCKS_PER_SEC;

  printf("Execution Time: %f seconds\n", execution_time);
  printf("Valleys: %d\n", result);

  return 0;
}
